Access Plan: WIP

eri-trabiccolo opened the pull request:
## Description
WIP for #49 
I mostly wanted to open this so that you could take a preliminary look at it @thomasplevy 
I followed and tried to interpretate the specs...
We're probably going to discuss a little more in depth some properties... 
I think we're probably going to not expose some `'yesno'` ones.

## How has this been tested?
not tested yet if not a rough manual one.


## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/master/tests/README.md -->
- [x] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/master/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/master/docs/documentation-standards.md -->

thomasplevy commented:
ooo exciting, I'll look closer this afternoon!

eri-trabiccolo committed:
wip

thomasplevy commented on spec/components/schemas/AccessPlan.yaml:
```diff
@@ -51,22 +51,23 @@ allOf:
         type: number
         format: float
         minimum: 0
         example: 1.00
```
The example object displays `1.00` as `1` which might be misleading at a glance.

```suggestion
        example: 99.97
```

thomasplevy commented on spec/components/schemas/AccessPlan.yaml:
```diff
@@ -92,7 +93,7 @@ allOf:
-          Only applicable when $access_expiration is `limited-date`.
+          Only applicable when `access_expiration` is `limited-date`.
           Format: `Y-m-d H:i:s`.
         type: string
```
Add example

```suggestion
        type: string
        example: '2019-03-23 23:59:59'
```

thomasplevy reviewed (commented):
Minor changes /  Notes on spec:

+ `post_id` property should be required, I see it is required on creation but I'm not seeing it required on the generated object.
+ Add a `<br>` to make this area of the spec more readable

![Screenshot_2021-01-04_10-25-02](https://user-images.githubusercontent.com/1290739/103561760-d614f980-4e6e-11eb-9681-a1e6b3f0b947.png)

+ The `GET` endpoint is showing a `password` query string variable. This isn't necessary since we don't implement the ability to password protect access plans.

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -80,14 +80,125 @@ function llms_rest_server_error( $message = '' ) {
+	if ( ! empty( $memberships ) ) {
+		$real_memberships = array_map(
+			function( $membership ) {
+				return is_a( llms_get_post( $membership ), 'LLMS_Membership' );
```
I have not run any benchmarks on this but I believe that using a WP Core method here will be slightly more performant than using our post model class. I might be wrong about this, though.

```suggestion
				return ( 'llms_membership' === get_post_type( $membership ) );
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -80,14 +80,125 @@ function llms_rest_server_error( $message = '' ) {
+ * @return boolean
+ */
+function llms_rest_validate_positive_float( $number, $include_zero = true ) {
+	// min_range and max_range options for FILTER_VALIDATE_FLOAT are only available since PHP 7.4.
```
One day we can find this note and make an update to use the filter

```suggestion
	// @TODO min_range and max_range options for FILTER_VALIDATE_FLOAT are only available since PHP 7.4.
```

thomasplevy commented:
I haven't looked at the main controller class yet

eri-trabiccolo committed:
Apply suggestions from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Apply suggestions from code review
+ fix occurrences where we returned a not defined var instead of a WP_Error preparing collection query args in various posts based resources
+ add access plan post_id collection query param
+ fix some spec typos

eri-trabiccolo commented:
@thomasplevy 
> Minor changes / Notes on spec:
> 
> * `post_id` property should be required, I see it is required on creation but I'm not seeing it required on the generated object.

I don't think I get this. What do you mean?

> * Add a `<br>` to make this area of the spec more readable
> 
> ![Screenshot_2021-01-04_10-25-02](https://user-images.githubusercontent.com/1290739/103561760-d614f980-4e6e-11eb-9681-a1e6b3f0b947.png)

Done, thanks!

> * The `GET` endpoint is showing a `password` query string variable. This isn't necessary since we don't implement the ability to password protect access plans.

Done, thanks!

eri-trabiccolo committed:
minor cleaning

eri-trabiccolo committed:
remove access plan post 'status' property

eri-trabiccolo committed:
ap: fix post_id collection param usage

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -461,8 +461,13 @@ public function create_item_permissions_check( $request ) {
-				return llms_rest_authorization_required_error( sprintf( __( 'Sorry, you are not allowed to create a %s as this user.', 'lifterlms' ), $post_type_object->labels->singular_name ) );
+				return llms_rest_authorization_required_error(
+					sprintf(
+						// Translators: The post type singular name.
+						__( 'Sorry, you are not allowed to create a %s as this user.', 'lifterlms' ),
+						$post_type_object->labels->singular_name
```
1. I like the `$var = $desc` pattern here, I've never actually seen it used before but I tend towards this myself. 
2. I realize that in normal scenarios (courses and memberships) this won't matter but just in case using the name (plural) without a will prevent potential grammatical issues with post types that start with a vowel.

```suggestion
						// Translators: %s = The post type name.
						__( 'Sorry, you are not allowed to create %s as this user.', 'lifterlms' ),
						$post_type_object->labels->name
```

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -493,11 +498,15 @@ public function update_item_permissions_check( $request ) {
-				return llms_rest_authorization_required_error( sprintf( __( 'Sorry, you are not allowed to update a %s as this user.', 'lifterlms' ), $post_type_object->labels->singular_name ) );
+				return llms_rest_authorization_required_error(
+					sprintf(
+						// Translators: The post type singular name.
+						__( 'Sorry, you are not allowed to update a %s as this user.', 'lifterlms' ),
+						$access_plan->get_post_type_label()
```
```suggestion
						// Translators: %s = The post type name.
						__( 'Sorry, you are not allowed to update %s as this user.', 'lifterlms' ),
						$access_plan->get_post_type_label( 'name' )
```

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -528,11 +537,15 @@ public function delete_item_permissions_check( $request ) {
-				return llms_rest_authorization_required_error( sprintf( __( 'Sorry, you are not allowed to delete a %s as this user.', 'lifterlms' ), $post_type_object->labels->singular_name ) );
+				return llms_rest_authorization_required_error(
+					sprintf(
+						// Translators: The post type singular name.
+						__( 'Sorry, you are not allowed to delete a %s as this user.', 'lifterlms' ),
+						$access_plan->get_post_type_label()
```
```suggestion
						// Translators: %s = The post type name.
						__( 'Sorry, you are not allowed to delete %s as this user.', 'lifterlms' ),
						$access_plan->get_post_type_label( 'name' )
```

thomasplevy reviewed (commented):
Minor (mostly grammatical) suggestions

eri-trabiccolo committed:
Apply suggestions from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Turn price-ish properties' type into number (float isn't allowed).
Initial tests.

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -629,7 +629,7 @@ protected function prepare_object_for_response( $access_plan, $request ) {
 
 		// Price.
-		$data['price'] = $access_plan->is_free() ? 0 : $access_plan->get_price( 'price', array(), 'float' );
+		$data['price'] = $access_plan->is_free() ? 0 : $access_plan->get_price( 'price', array(), 'number' );
```
The method doesn't have a `number` option. Using this defaults to no formatting which is perhaps what you're desiring but I'm not sure why

eri-trabiccolo replied:
No that's just a bad search and replace :(

thomasplevy replied:
Ah, glad I noticed it in that case!

eri-trabiccolo replied:
fixed in https://github.com/gocodebox/lifterlms-rest/pull/215/commits/ed4f6b0d5e0b1375052633000564fa20112d7cdd

eri-trabiccolo committed:
fix getting access plan price with wrong formatting option

eri-trabiccolo committed:
add some access plan manipulation permissions tests

eri-trabiccolo committed:
Add further access-plans tests
also introduce helpers to determine the nature of the wp errors used in rest routines

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -26,12 +27,12 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
 		$error_code = 'llms_rest_forbidden_request';
 		$_message   = __( 'You are not authorized to perform this request.', 'lifterlms' );
-		$status     = '403';
+		$status     = WP_Http::FORBIDDEN;
```
```suggestion
		$status     = WP_Http::FORBIDDEN; // 403.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -26,12 +27,12 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
 		$error_code = 'llms_rest_unauthorized_request';
 		$_message   = __( 'The API credentials were invalid.', 'lifterlms' );
-		$status     = '401';
+		$status     = WP_Http::UNAUTHORIZED;
```
```suggestion
		$status     = WP_Http::UNAUTHORIZED; // 401.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
 function llms_rest_not_found_error( $message = '' ) {
 	$message = ! $message ? __( 'The requested resource could not be found.', 'lifterlms' ) : $message;
-	return new WP_Error( 'llms_rest_not_found', $message, array( 'status' => 404 ) );
+	return new WP_Error( 'llms_rest_not_found', $message, array( 'status' => WP_Http::NOT_FOUND ) ); // 404
```
```suggestion
	return new WP_Error( 'llms_rest_not_found', $message, array( 'status' => WP_Http::NOT_FOUND ) ); // 404.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ * @since [version]
+ *
+ * @param WP_Error $wp_error The WP_Error object.
+ * @return array
```
This will be an array of error code integers, right? EG: `array( 404, 403 )` or something, right?

```suggestion
 * @return int[]
```

thomasplevy reviewed (commented):
This isn't a full review, just some small doc things I noticed while keeping an eye on the progress

eri-trabiccolo committed:
Apply suggestions from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Better handling of properties interdependency

eri-trabiccolo committed:
Add further tests

eri-trabiccolo committed:
add some tests

eri-trabiccolo commented:
@thomasplevy I've added a bunch of tests, cannot say it's full-covered though... I think it's fine.

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ * @return boolean
+ */
+function llms_rest_is_bad_request_error( $wp_error ) {
+	return in_array( WP_Http::BAD_REQUEST, llms_rest_get_all_error_statuses( $wp_error ), true );
```
I'm adding notes here because I can tell I'm going to be constantly looking up what the numbers actually are because. Most of them are easy to remember but I can almost never rember the difference between 401 and 403, primarily.

Having the comment I think helps... maybe I'm alone in this memory issue but... well, I work with the code so I want it there <3

```suggestion
	return in_array( WP_Http::BAD_REQUEST, llms_rest_get_all_error_statuses( $wp_error ), true ); // 400.
```

eri-trabiccolo replied:
You're right, and you're not alone.

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ * @return boolean
+ */
+function llms_rest_is_authorization_required_error( $wp_error ) {
+	return ! empty( array_intersect( llms_rest_get_all_error_statuses( $wp_error ), array( WP_Http::FORBIDDEN, WP_Http::UNAUTHORIZED ) ) );
```
```suggestion
	return ! empty( array_intersect( llms_rest_get_all_error_statuses( $wp_error ), array( WP_Http::FORBIDDEN, WP_Http::UNAUTHORIZED ) ) ); // 403, 401.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ * @return boolean
+ */
+function llms_rest_is_not_found_error( $wp_error ) {
+	return in_array( WP_Http::NOT_FOUND, llms_rest_get_all_error_statuses( $wp_error ), true );
```
```suggestion
	return in_array( WP_Http::NOT_FOUND, llms_rest_get_all_error_statuses( $wp_error ), true ); // 404.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+}
+
+/**
+ * Checks whether or not the passed object is a not found error
```
```suggestion
 * Checks whether or not the passed object is a 404 not found error
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+}
+
+/**
+ * Checks whether or not the passed object is an internal server error
```
```suggestion
 * Checks whether or not the passed object is a 500 internal server error
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ * @return boolean
+ */
+function llms_rest_is_server_error( $wp_error ) {
+	return in_array( WP_Http::INTERNAL_SERVER_ERROR, llms_rest_get_all_error_statuses( $wp_error ), true );
```
```suggestion
	return in_array( WP_Http::INTERNAL_SERVER_ERROR, llms_rest_get_all_error_statuses( $wp_error ), true ); // 500.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+}
+
+/**
+ * Checks whether or not the passed object is a bad request error
```
```suggestion
 * Checks whether or not the passed object is a 400 bad request error
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
 
 /**
- * Validate submitted array of integers is an array of real real user ids.
+ * Checks whether or not the passed object is an error of missing required authorization
```
```suggestion
 * Checks whether or not the passed object is a 401 (permission) or 403 (authorization) error
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ *
+ * @since [version]
+ *
+ * @param integer $number Integer number to validate.
+ * @param boolean $include_zero Optional. Whether or not 0 is included. Default is `true`.
```
```suggestion
 * @param integer $number       Integer number to validate.
 * @param boolean $include_zero Optional. Whether or not 0 is included. Default is `true`.
```

thomasplevy commented on includes/server/llms-rest-server-functions.php:
```diff
@@ -42,52 +43,250 @@ function llms_rest_authorization_required_error( $message = '', $check_authentic
+ *
+ * @since [versoin]
+ *
+ * @param integer $number Float number to validate.
+ * @param boolean $include_zero Optional. Whether or not 0 is included. Default is `true`.
```
```suggestion
 * @param integer $number       Float number to validate.
 * @param boolean $include_zero Optional. Whether or not 0 is included. Default is `true`.
```

thomasplevy commented:
@eri-trabiccolo reviewing now

eri-trabiccolo committed:
remove stray comma

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -0,0 +1,1081 @@
+	 */
+	public function create_item_permissions_check( $request ) {
+
+		$can_create = parent::create_item_permissions_check( $request );
+
+		// If current user cannot create the item because of authorization, check if the current user can edit the "parent" course/membership.
+		if ( llms_rest_is_authorization_required_error( $can_create ) ) {
```
I think we can probably skip the parent method check and always rely on this parent (course/membership) post check below.

I think, right? What do we gain by doing both?

What do we gain by doing both checks?

eri-trabiccolo replied:
Well the idea is that we want to keep _all_ the checks we make on the abstract, which atm aren't that many, but stil that's the idea of abstraction... I think eheh
E.g.
https://github.com/gocodebox/lifterlms-rest/blob/1.0.0-beta.17/includes/abstracts/class-llms-rest-posts-controller.php#L187-L190
If I don't call the parent class method I have to do that check here.
Also, in the parent class method we also check if the current user can create this post type (publish), which for _admins_ is always true: for them we don't need to check if they can edit the parent course/membership.

I know that _practically_ for _admins_ checking if they can publish access plans is equivalent to check if they can edit the parent course, but still what we're doing here is _formally_ more correct.
I'd like to keep this as it is :D
But let me know if you prefer otherwise.

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -0,0 +1,1081 @@
+				return $can_update;
+			}
+
+			$product_id               = $access_plan->get( 'product_id' );
+			$product_post_type_object = get_post_type_object( get_post_type( $product_id ) );
+
+			if ( ! current_user_can( $product_post_type_object->cap->edit_post, $product_id ) ) {
+				return llms_rest_authorization_required_error(
+					sprintf(
+						// Translators: %s = The post type name.
+						__( 'Sorry, you are not allowed to update %s as this user.', 'lifterlms' ),
+						$access_plan->get_post_type_label( 'name' )
+					)
+				);
+			}
+
+			$can_update = true;
+		}
```
This particular bit is repeated for the other permission checks, I think it would be good to move this check into a protected method so we can reuse it instead of repeating it.

eri-trabiccolo replied:
Right, refactored a little bit.

thomasplevy commented on includes/server/class-llms-rest-access-plans-controller.php:
```diff
@@ -0,0 +1,1081 @@
+		$product_id = $access_plan->get( 'product_id' );
+		$post       = llms_get_post( $product_id );
+
+		// Post: the check below should be redundant but at the moment we don't delete
+		// access plans when deleting "related" courses/memberships.
+		if ( $post instanceof LLMS_Course || $post instanceof LLMS_Membership ) {
+			$links['post'] = array(
+				'href' => rest_url(
+					sprintf(
+						'%s/%s/%s',
+						'llms/v1',
+						$access_plan->get_product_type(),
+						$product_id
+					)
+				),
+			);
+		}
```
I don't like having a hack in here for a bug that we can fix in the core... I think we should remove this and just fix the core bug simultaneously: https://github.com/gocodebox/lifterlms/issues/263

thomasplevy replied:
That said... I'm okay with shipping it if we open a new issue (in rest) so that we can make sure we don't forget about it when we update the core

eri-trabiccolo replied:
Yeah but unless we also update the db removing orphans access plans, the check here is needed.
So yeah, let's also update the db.

thomasplevy replied:
okay sounds good

eri-trabiccolo replied:
I'm on this right now (db update).
Did the ap deletion on course/membership deletion.

eri-trabiccolo replied:
ap clean up in core, see https://github.com/gocodebox/lifterlms/pull/1518

eri-trabiccolo committed:
Apply suggestions from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Improve some error messages

eri-trabiccolo committed:
doc fix

eri-trabiccolo committed:
Refactor access plan permissions check

thomasplevy commented on tests/unit-tests/server/class-llms-rest-test-access-plans.php:
```diff
@@ -0,0 +1,868 @@
+
+		$response = $this->perform_mock_request(
+			'DELETE',
+			$this->route . '/12569',
```
```suggestion
			$this->route . '/12569'
```

eri-trabiccolo committed:
remove stray comma

eri-trabiccolo committed:
remove check on whether the access plan product exists - we remove orphan aps in core

eri-trabiccolo committed:
Fix access plans links

eri-trabiccolo commented:
@thomasplevy the spec I used didn't expose the `post_type` property, which actually we use to expose for all the other post based resources. Do you think we should?

thomasplevy commented:
@eri-trabiccolo yea, for consistency I think we should. I'm going to merge this and I'll add the prop to the spec on the dev branch

eri-trabiccolo commented:
> @eri-trabiccolo yea, for consistency I think we should. I'm going to merge this and I'll add the prop to the spec on the dev branch

@thomasplevy cool, but you need to update the resource and the tests as well :D
https://github.com/gocodebox/lifterlms-rest/pull/221/files#diff-13755cbf4d8d2058eef40059fd2ba3d8e36aae15e8026be127b161fd81c828ddR53

and
https://github.com/gocodebox/lifterlms-rest/pull/221/files#diff-e819f7b86cf85964c41377a62f616c72f1eeacec99a495b23192944dac423676R51

thomasplevy commented:
Alright, I didn't think about that, I already updated the spec on dev
